app/src/gui/components/map/tile-source.ts

```diff
@@ -0,0 +1,588 @@
+    if (!this.db.objectStoreNames.contains(this.dbName)) {
+      this.db.createObjectStore(this.dbName, {
+        keyPath: this.keyPath,
+      });
+    }
+  }
+
+  async get(query: IDBKeyRange | IDBValidKey): Promise<Type | undefined> {
+    return new Promise<Type | undefined>((resolve, reject) => {
+      if (!this.db) {
+        return;
+      }
+      const transaction = this.db.transaction(this.dbName, 'readonly');
+      const store = transaction.objectStore(this.dbName);
+      const request = store.get(query);
+      request.onsuccess = () => {
+        resolve(request.result);
+      };
+      request.onerror = () => reject(request.error);
+    });
+  }
+
+  async getAll(): Promise<Type[] | undefined> {
+    return new Promise<Type[] | undefined>((resolve, reject) => {
+      if (!this.db) {
+        return;
+      }
+      const transaction = this.db.transaction(this.dbName, 'readonly');
+      const store = transaction.objectStore(this.dbName);
+      const request = store.getAll();
+      request.onsuccess = () => {
+        resolve(request.result);
+      };
+      request.onerror = () => reject(request.error);
+    });
+  }
+
+  async put(object: Type) {
+    return new Promise<IDBValidKey | undefined>((resolve, reject) => {
+      if (!this.db) {
+        return;
+      }
+      const transaction = this.db.transaction(this.dbName, 'readwrite');
+      const store = transaction.objectStore(this.dbName);
+      const request = store.put(object);
+      request.onsuccess = () => resolve(request.result);
+      request.onerror = () => reject(request.error);
+    });
+  }
+
+  async delete(key: IDBValidKey) {
+    return new Promise<void>((resolve, reject) => {
+      if (!this.db) {
+        return;
+      }
+      const transaction = this.db.transaction(this.dbName, 'readwrite');
+      const store = transaction.objectStore(this.dbName);
+      const request = store.delete(key);
+      request.onsuccess = () => resolve();
+      request.onerror = () => reject(request.error);
+    });
+  }
+
+  async clear() {
+    if (this.db) {
+      return new Promise<void>((resolve, reject) => {
+        const transaction = this.db.transaction(this.dbName, 'readwrite');
+        const store = transaction.objectStore(this.dbName);
+        transaction.oncomplete = () => resolve();
+        transaction.onerror = () => reject(transaction.error);
+        store.clear();
+      });
+    }
+  }
+}
+
+class TileStoreBase {
+  // The database is a static member of this class, there is only
+  // one connection to the DB in the app
+  static DB_NAME = 'tiles_db';
+  static db: IDBDatabase;
+  // references to the individual object stores within the database
+  tileDB!: IDB<StoredTile>;
+  tileSetDB!: IDB<StoredTileSet>;
+
+  constructor() {
+    this.initDB();
+  }
+
+  // initDB is called from the constructor but clients may want
+  // to call it directly and wait for it to resolve if they
+  // need to ensure that the databases are ready
```

Comment:
Can you tie in initDB into the initialisation logic which already exists in the app? Then it can be awaited as the app is loading